### PR TITLE
Improvements to `run` and `list` commands

### PR DESCRIPTION
**Pull request by christian-bromann**
According to our internal proposal this patch makes some changes to the `run` and `list` commands. These features are behind an environment flag `RUNME_EXPERIMENTAL_CLI`. So the current behavior doesn't change. The new (experimental) CLI behavior works as following:

- the `list` command now lists all commands of all markdown files within a project
  - it only shows the command id and the description, everything else seems not relevant
- the `run` command
  - now only takes the block name to execute it
  - can receive more block names which will be run sequentially
  - now has a `-p` or `--parallel` flag to run a set of commands in parallel
  - allows to execute sequential commands and parallel ones through one CLI call, e.g. `runme run foo bar loo -p faz baz -s loo bar` will run: `foo`, `bar` and `loo` sequentially, then will run `faz` and `baz` in parallel and lastly `loo` and `bar` sequential again

__List Command Demo:__
Screenshot 2023-05-01 at 22 59 55

Screenshot of the WebdriverIO project based on [`webdriverio/webdriverio#10237`](https://github.com/webdriverio/webdriverio/pull/10237)

__Run Command Demo:__
run

__Reviewer Notes:__
This patch basically does the following:
- moves flat functions like `getCodeBlocks` or `readMarkdownFile` into the project module
- implements a `getAllMarkdownFiles` in the project module to find all markdown files
- implement `LookUpCodeBlockById` to find a block just by its name
- __unrelated:__ removed `-s` for server address as it seems to be better suited (and more used) for providing the sequential flag

__Outstanding:__
- [x] read from `.gitignore` to not include markdown files matching these patterns (otherwise we would e.g. read from a bunch of `node_modules` which would be quite overwhelming
- [ ] write unit tests
- [ ] more thorough testing of new functionality
- [ ] detect duplicate commands and throw errors
- [x] improve error handling

### CODE REVIEW

**Comment by sourishkrout** [comment]
The scope of the original ticket https://github.com/stateful/runme/issues/229 is about changes, there wasn't green light for any other changes yet @christian-bromann.

File: internal/cmd/root.go [hunk 1]
```diff
@@ -47,9 +49,11 @@ func Root() *cobra.Command {
 	pflags := cmd.PersistentFlags()
 
 	pflags.BoolVar(&fAllowUnknown, "allow-unknown", true, "Display snippets without known executor")
+	pflags.BoolVar(&fIgnoreNameless, "ignore-nameless", true, "Ignore code blocks without name defined")
```
Comment (mxsdev) [inline internal/cmd/root.go:52]:
```suggestion
	pflags.BoolVar(&fIgnoreNameless, "ignore-nameless", false, "Ignore code blocks without name defined")
```

Should probably be false by default.

File: internal/project/project.go [hunk 2]
```diff
@@ -1,241 +1,232 @@
...
-func (r *repo) Err() error { return r.err }
-
-func (r *repo) BranchName() string {
-	if r.Err() != nil {
-		return ""
-	}
-
-	ref, err := r.Head()
+func (p *Project) getAllMarkdownFiles() ([]string, error) {
+	root, err := os.Getwd()
```
Comment (mxsdev) [inline internal/project/project.go:60]:
```suggestion
	root := p.RootDir
```

File: internal/project/project.go [hunk 3]
```diff
@@ -1,241 +1,232 @@
...
 	if err != nil {
-		r.err = err
-		return ""
+		return nil, err
 	}
 
-	if ref.Name().IsBranch() {
-		return ref.Name().Short()
+	fsys := os.DirFS(root)
+	matches, err := fs.Glob(fsys, "**/*.md")
```
Comment (mxsdev) [inline internal/project/project.go:66]: This [does not work](https://github.com/golang/go/issues/11862) as expected, and in particular will ignore files in the root directory.

You should probably use a `WalkDir`-based solution instead.

File: internal/project/project.go [hunk 4]
```diff
@@ -1,241 +1,232 @@
...
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	branches, err := GetBranchNames(cwd)
-	if err != nil {
-		return nil, errors.New("error while querying repository branches")
+	for _, file := range files {
+		codeBlock, err := p.GetCodeBlocks(file[len(p.RootDir):], false, true)
```
Comment (mxsdev) [inline internal/project/project.go:219]:
```suggestion
		codeBlock, err := p.GetCodeBlocks(file[len(p.RootDir):], allowUnknown, ignoreNameless)
```

Should pass these down

File: internal/cmd/run.go [hunk 5]
```diff
@@ -3,126 +3,344 @@ package cmd
...
+
+	var stdin io.Reader
+
+	if isInExperimentalMode() && block.Interactive() {
+		warn := color.New(color.Bold, color.FgYellow)
+		_, err := warn.Print("Warning: ")
+		if err != nil {
+			return err
+		}
+		fmt.Println("Interactive mode was disabled as it is not supported with the experimental CLI")
```
Comment (mxsdev) [inline internal/cmd/run.go:245]:
```suggestion
		fmt.Println("Interactivity was disabled as it is not supported with the experimental CLI")
```

File: internal/cmd/run.go [hunk 6]
```diff
@@ -3,126 +3,344 @@ package cmd
...
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		// carry on
+		break
```
Comment (mxsdev) [inline internal/cmd/run.go:337]: This break does nothing since there is no surrounding loop

**Review by mxsdev — CHANGES_REQUESTED** [review]
Really loving the UX changes, would love to see this implemented asap!

The runner-based implementation is also reasonable, though I do have some changes in mind. But the broad-strokes idea is good!

File: internal/cmd/run.go [hunk 7]
```diff
@@ -3,126 +3,344 @@ package cmd
...
-			}
+	setDefaultFlags(&cmd)
+
+	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "Print the final command without executing.")
+	cmd.Flags().StringArrayVarP(&replaceScripts, "replace", "r", nil, "Replace instructions using sed.")
+
+	if isInExperimentalMode() {
+		cmd.Flags().BoolVarP(&runInParallel, "parallel", "p", false, "Run commands in parallel. (experimental)")
+		cmd.Flags().BoolVarP(&runSequential, "sequential", "s", true, "Run commands sequentially. (experimental)")
+		cmd.Flags().BoolVarP(&onlyCommandIO, "onlyCommandOutput", "", false, "If set, Runme will only output command output. (experimental)")
```
[the remark this answers is not in the document]
Comment (christian-bromann) [reply internal/cmd/run.go:155]: Oh yeah totally! Nice

**Comment by christian-bromann** [comment]
Closing in favor of #236

**Pull request closed without merging**